dvc/ignore.py

```diff
@@ -75,18 +81,29 @@ class DvcIgnoreFilter(object):
     def __init__(self, root_dir, tree):
         self.tree = tree
         self.ignores = {DvcIgnoreDirs([".git", ".hg", ".dvc"])}
-        self._update(root_dir)
-        for root, dirs, _ in self.tree.walk(root_dir, dvcignore=self):
-            for d in dirs:
-                self._update(os.path.join(root, d))
 
     def _update(self, dirname):
         ignore_file_path = os.path.join(dirname, DvcIgnore.DVCIGNORE_FILE)
         if self.tree.exists(ignore_file_path):
             self.ignores.add(DvcIgnorePatterns(ignore_file_path, self.tree))
 
+    def _update_parents(self, path):
+        if not os.path.isabs(path):
+            path = os.path.join(self.tree.tree_root, path)
+        while path != self.tree.tree_root:
+            self._update(path)
+            prev_path = path
+            path = os.path.dirname(path)
+            if prev_path == path:
+                break
+        self._update(path)
```

Comment:
Assuming that this method stays, I do realize that handling of this while loop should probably be improved.